peer.go

```diff
@@ -0,0 +1,1087 @@
+	controlCh := make(chan peerControl)
+	p := makePeerInternal(serverID, serverAddress, logs, snapshots,
+		goRoutines, trans, localAddr, controlCh, progressCh, options)
+	p.shared.goRoutines.spawn(p.selectLoop)
+	return controlCh
+}
+
+func makePeerInternal(serverID ServerID,
+	serverAddress ServerAddress,
+	logs LogStore,
+	snapshots SnapshotStore,
+	goRoutines *waitGroup,
+	trans Transport,
+	localAddr ServerAddress,
+	controlCh <-chan peerControl,
+	progressCh chan<- peerProgress,
+	options peerOptions) *peerState {
+
+	if options.logger == nil {
+		options.logger = log.New(os.Stderr, "", log.LstdFlags)
+	}
+	if options.maxAppendEntries == 0 {
+		options.maxAppendEntries = 1000
+	}
+	if options.heartbeatInterval == 0 {
+		options.heartbeatInterval = 100 * time.Millisecond
+	}
+	if options.failureWait == 0 {
+		options.failureWait = 10 * time.Millisecond
+	}
+	if options.maxFailureWait == 0 {
+		options.maxFailureWait = 100 * time.Millisecond
+	}
+
+	p := &peerState{
+		shared: &peerShared{
+			options:            options,
+			peerID:             serverID,
+			peerAddr:           serverAddress,
+			trans:              trans,
+			localAddr:          localAddr,
+			logger:             options.logger,
+			logs:               logs,
+			snapshots:          snapshots,
+			goRoutines:         goRoutines,
+			requestCh:          make(chan peerRPC),
+			replyCh:            make(chan peerRPC),
+			pipelineSendDoneCh: make(chan struct{}),
+			stopCh:             make(chan struct{}),
+		},
+		controlCh:  controlCh,
+		progressCh: progressCh,
+
+		backoffTimer: time.NewTimer(time.Hour),
+		progress: peerProgress{
+			peerID: serverID,
+		},
+	}
+	p.backoffTimer.Stop()
+
+	pipeline, err := p.shared.trans.AppendEntriesPipeline(p.shared.peerAddr)
+	if err != nil {
+		if err != ErrPipelineReplicationNotSupported {
+			p.shared.logger.Printf("[ERR] raft: Failed to start pipeline replication to %s: %s",
+				p.shared.peerAddr, err)
+		}
+	} else {
+		p.shared.pipeline = pipeline
+	}
+
+	return p
+}
+
+func (p *peerState) checkInvariants() error {
+	if p.shared.peerID != p.progress.peerID {
+		return fmt.Errorf("Progress must have peerID")
+	}
+	if p.progress.lastContact.After(p.progress.lastReply) ||
+		p.progress.lastContact == p.progress.lastReply && !p.progress.lastContact.IsZero() {
+		return fmt.Errorf("lastContact must be after lastReply (or both 0)")
+	}
+	if p.nextIndex > p.control.lastIndex+1 {
+		return fmt.Errorf("nextIndex must not be more than 1 entry past the lastIndex")
+	}
+	if p.outstandingInstallSnapshotRPC && p.outstandingAppendEntriesRPCs > 0 {
+		return fmt.Errorf("must not send a snapshot and entries simultaneously")
+	}
+	return nil
+}
+
+// selectLoop is a long-running routine that sends RPCs to a single remote
+// server.
+func (p *peerState) selectLoop() {
+	for {
+		// At this point, we may or may not have more work that we could do right away.
+		// Check for new information but don't block.
+		p.drainNonblocking()
```

Comment:
@slackpad, what I was concerned about was, using your proposed loop structure:

1. Blocking select: get an RPC reply that enables a snapshot to be sent.
2. Poll control: get a new verifyCounter that enables a heartbeat to be sent.
3. issueWork: sends heartbeat.
4. Blocking select: blocks.

We should have sent that snapshot instead of blocking. Oops.

And in general, it'd be nice if a single receive can enable multiple actions to fire. This leads to a structure like:

```
for {
  if issueWork() {
    continue
  }
  blockingSelect()
}
```

But then the concern is that blockingSelect() has no priority, so this might result in suboptimal schedules. For example, it might receive an RPC reply rather than new control information with a new term number. If we acted immediately, we might send an RPC that we could have avoided by polling. Are these suboptimal schedules a big deal? Shrug.

So do we poll just the control channel but block on all the channels? Which ones are really important to justify polling on? The channels are these:
 1. recv control
 2. send progress
 3. recv prepared RPC request
 4. recv RPC reply
 5. recv pipeline send completed
 6. recv backoff timer expired
 7. recv heartbeat timer expired (you can't poll this the way the code is structured now)

I think that neglecting (1) and (4) and possibly (3) could result in sending suboptimal RPC requests.

Then it's a judgment call:
 - duplicate none (my for loop above): It's simplest. We'll embrace any performance hit until we learn better.
 - duplicate all (the original patch): It's mechanical. There's no tough decision to make about whether a channel goes in both the poll and the blocking select: they all do.
 - duplicate some: If we pick the right subset, we should get the performance benefits of duplicating all but with less code duplication.